ansible/roles/test/files/ptftests/advanced-reboot.py

```diff
@@ -947,31 +976,131 @@ def check_alive(self):
-            if success:
-              was_alive = True
+            state = self.get_asic_state()
+            if state == 'up':
+                if not uptime:
+                    uptime = self.get_asic_state_time(state)
             else:
-              if was_alive:
-                return False    # Stopped working after it working for sometime?
+                if uptime:
+                    return False # Stopped working after it working for sometime?
+            time.sleep(2)
 
         # wait, until FDB entries are populated
         for _ in range(self.nr_tests * 10): # wait for some time
-            if not self.ping_alive()[1]:    # until we see that there're no extra replies
+            if not self.asic_flooding:
                 return True
+            time.sleep(2)
 
         return False                        # we still see extra replies
 
-    def ping_alive(self):
-        nr_from_s = self.pingFromServers()
-        nr_from_l = self.pingFromUpperTier()
 
-        is_alive      = nr_from_s > self.nr_pc_pkts * 0.7 and nr_from_l > self.nr_vl_pkts * 0.7
-        is_asic_weird = nr_from_s > self.nr_pc_pkts        or nr_from_l > self.nr_vl_pkts
-        # we receive more, then sent. not populated FDB table
+    def get_asic_state(self):
+        with self.asic_state_lock:
+            state = self.asic_state
+        return state
+
+
+    def get_asic_state_time(self, state):
+        with self.asic_state_lock:
+            time = self.asic_state_time[state]
+        return time
+
+
+    def set_asic_state(self, state, reachability=0):
+        with self.asic_state_lock:
+            self.asic_state             = state
+            self.asic_state_time[state] = datetime.datetime.now()
+
+
+    def get_asic_vlan_reachability(self):
+        return self.asic_vlan_reach
+
+
+    def asic_start_recording_vlan_reachability(self):
+        with self.asic_state_lock:
+            self.asic_vlan_reach = []
+            self.recording       = True
+
+
+    def asic_stop_recording_vlan_reachability(self):
+        with self.asic_state_lock:
+            self.recording = False
+
+
+    def try_record_asic_vlan_recachability(self, t1_to_vlan):
+        with self.asic_state_lock:
+            if self.recording:
+                self.asic_vlan_reach.append(t1_to_vlan)
+
+    def get_cpu_state(self):
+        with self.cpu_state_lock:
+            state = self.cpu_state
+        return state
+
+
+    def get_cpu_state_time(self, state):
+        with self.cpu_state_lock:
+            time = self.cpu_state_time[state]
+        return time
+
+
+    def set_cpu_state(self, state):
+        with self.cpu_state_lock:
+            self.cpu_state             = state
+            self.cpu_state_time[state] = datetime.datetime.now()
+
+
+    def log_asic_state_change(self, reachable, partial=False, t1_to_vlan=0):
+        old = self.get_asic_state()
+
+        if reachable:
+            state = 'up' if not partial else 'partial'
```

Comment:
How we're going to treat partial? Good or not?

Reply:
Right now, it is for state transition record. I didn't change logic, so partial is treated as 'up' at the moment. I am adding this state so that we could use it in the future if we choose to.